SPEC.md

```diff
@@ -0,0 +1,121 @@
+In this case, anytime the `#` is a significant character in the core language, the `;` character should be as well.
+
+## Extensions
+All extensions are fully optional.
+Currently, multiple of them are consequences of keeping the grammar lean, and are documented, but only parts of their behaviors are desirable.
+
+### Flexspace
+In implementations with the "flexspace" extension, whitespace within statements is not significant.
+This means that you may have arbitrary whitespace (not only spaces, but also tabs, newlines, zero-width spaces, and so on) in the following locations:
+* Between the opening section bracket, the section name, and the closing section bracket.
+* Between the key in a key-(raw)value pair, the `=`, and the (raw)value.
+* Between a (raw) value and a comment.
+
+### Tabulation
+In implementations with the "tabulation" extension, whitespace before and after statements is not significant.
+This means you may have arbitrary whitespace in the following locations:
+* In front of a section declaration.
+* In front of a key-(raw)value pair.
+* In front of a comment.
+* Between a key-rawvalue pair and the next statement.
+* Between a section declaration and the next statement.
+
+## API
+This API is purely a suggestion for the sake of users - implementations may follow any scheme they wish.
+
+Backend storage should be implemented in a trie-like fashion, separated by `.`s.
+It should be possible to get the direct list of children of a trie node, as well as the full list of children of a trie node.
+It should be possible to know if a given key is a section, key, both, or neither.
+It should be possible to walk over matches without requiring allocation of a new trie.
+Keys do not need to be ordered in any specific way, but sorting them may make implementing the API easier.
+
+The following API endpoints provide all of this functionality:
+
+### Walk (Flat|Rec)
+WalkFlat|WalkRec takes a pattern (which may be empty) and a function.
+It will call the function with each matching key-value pair.
+The matches are performed like so:
+1. If the pattern is empty, all keys match.
+2. If the pattern is an invalid key, no keys match.
+3. If the pattern is a valid key:
+   With the Rec variant, all keys that start with the pattern followed by a "." match.
+   With the Flat variant, all keys that start with the pattern followed by a ".", but that do not contain any additional "."s beyond those match.
+
+### List (Flat|Rec)
+ListFlat|ListRec takes a pattern and returns a list of values.
+The list will contain all of the values (including duplicates) that match under the corresponding `Walk*` function.
+It can be implemented in pseudocode using the `Walk*` functions like so:
+
+```
+ListFlat (pattern):
+	var output_list
+	WalkFlat (pattern, (key, value){
+		output_list.append(value)
+	})
+	return output_list
+ListRec (pattern):
+	var output_list
+	WalkRec (pattern, (key, value){
+		output_list.append(value)
+	})
+	return output_list
```

Comment:
The syntax of this pseudocode seems a bit strange to me, especially the anonymous function syntax. It's a bit difficult to convey an advanced concept like a closure in pseudocode, and I don't have a better idea (yet).

Reply:
Yeah, it'd be nice to have some conventions for pseudocode, but as of right now we don't have any and this is (hopefully) clear enough.

Reply:
well we could look at what ~~Pseudocode~~ Python does
```suggestion
ListFlat (pattern):
	var output_list
	WalkFlat (pattern, lambda key, value:
		output_list.append(value)
	)
	return output_list
ListRec (pattern):
	var output_list
	WalkRec (pattern, lambda key, value:
		output_list.append(value)
	)
	return output_list
```
looks a bit more concise to me, but this much 🤏